src/ess.jl

```diff
@@ -0,0 +1,353 @@
+    # check arguments
+    niter = cache.niter
+    0 ≤ k < niter || throw(ArgumentError("only lags ≥ 0 and < $niter are supported"))
+
+    # compute mean autocorrelation
+    A = cache.A
+    nchains = size(A, 2)
+    s = zero(real(eltype(A)))
+    @inbounds for i in 1:nchains
+        s += real(A[k + 1, i]) / real(A[1, i])
+    end
+
+    return s / nchains
+end
+
+function mean_autocorr(k::Int, cache::BDAESSCache)
+    # check arguments
+    samples = cache.samples
+    niter, nchains = size(samples)
+    0 ≤ k < niter || throw(ArgumentError("only lags ≥ 0 and < $niter are supported"))
+
+    # compute mean autocorrelation
+	var = cache.var
+    s = zero(eltype(var))
+    @inbounds for j in 1:nchains
+        sj = zero(s)
+        for i in 1:(niter - k)
+            sj += abs2(samples[i, j] - samples[k + i, j])
+        end
+        s += sj / var[j]
+    end
+
+    return 1 - s / (2 * length(samples))
+end
+
+"""
+    ess(chains::Chains; sections = _default_sections(chains), kwargs...)
+
+Estimate the effective sample size and the potential scale reduction.
+"""
+function ess(
+    chains::Chains;
+    sections = _default_sections(chains), # perhaps directly nothing by default - or no sections argument?
+    kwargs...
+)
+    # subset the chain
+    _chains = Chains(chains, _clean_sections(chains, sections)) # may be unnecessary, already applied upstream
+
+    # estimate the effective sample size and rhat
+    ess, rhat = ess_rhat(_chains.value.data; kwargs...)
+
+    # convert to namedtuple
+    nt = merge((parameters = names(_chains),), (ess = ess, rhat = rhat))
+
+    return ChainDataFrame("ESS", nt)
+end
+
+function ess_rhat(chains::AbstractArray{<:Union{Missing,Real},3};
+                  method::AbstractESSMethod = ESSMethod(), kwargs...)
+    return ess_rhat(method, chains; kwargs...)
+end
+
+function ess_rhat(method, chains::AbstractArray{<:Union{Missing,Real},3};
+                  maxlag::Int = 250)
+    # compute size of matrices (each chain is split!)
+    niter = size(chains, 1) ÷ 2
+    nparams = size(chains, 2)
+    nchains = 2 * size(chains, 3)
+    ntotal = niter * nchains
+
+    # do not compute estimates if there is only one sample or lag
+    maxlag = min(maxlag, niter - 1)
+    maxlag > 0 || return fill(missing, nparams), fill(missing, nparams)
+
+    # define caches for mean and variance
+    U = typeof(float(zero(eltype(chains))))
+    T = eltype(chains) >: Missing ? Union{Missing,U} : U
+    chain_mean = Array{T}(undef, 1, nchains)
+    chain_var = Array{T}(undef, nchains)
+    samples = Array{T}(undef, niter, nchains)
+
+    # compute correction factor
+    correctionfactor = niter / (niter - 1)
+
+    # define cache for the computation of the autocorrelation
+    esscache = build_cache(method, samples, chain_var)
+ 
+    # define output arrays
+    ess = Vector{T}(undef, nparams)
+    rhat = Vector{T}(undef, nparams)
+
+    # for each parameter
+    for (i, chains_slice) in enumerate(eachslice(chains; dims = 2))
```

Comment:
My fault but I think
```suggestion
    for (i, chains_slice) in enumerate(eachcol(chains))
```
is easier to read.